### ISSUE TITLE
Improve "variations" detection by allowing different variable names

### ISSUE THREAD

**Issue by anter69** [Codewars/codewars.com#889]
In many (mostly easy) katas there are multiple "variations" of the same solution, which are not merged, because they use different variable names, e.g.

```python
def f(a, b):
    x = a + b
    return x
```
should be considered the "same" as 

```python
def f(a, b):
    res = a + b
    return res
```
and also

```python
def f(x, y):
    res = x + y
    return res
```

I'm sure it's not so trivial as detecting whitespace and different quote signs, but it would improve the "solutions" section a  bit, and it would also allow people getting some extra honor for "best practice", even if they used a different naming.

**Comment by brendanvos** [comment]
Want to work on that? As far as I can tell the code grouper is still located here https://github.com/Codewars/code-grouper.

**Comment by kazk** [comment]
I don't think `Codewars/code-grouper` is used, at least most the code seems unused.

> `similar?` and the related `difference` code is not currently being used on Codewars. Unless someone finds a good use for it, it can likely just be removed.

I just skimmed through `lib/code_grouper.rb` and most of the code looks related to `difference`. 
Also `same?` compares `reduced` code, but `strip_comments` only handles `ruby`/`javascript`/`coffeescript`. The last `push` was `2014-10-02` and many languages were added since then.

---

The "Solutions" section will definitely improve being uncluttered,  but I don't think people will get extra honor because only the user who submitted the first one of the variations gets actual honor points (the number on the profile should increase). See #147.

**Comment by jhoffner** [comment]
> I don't think Codewars/code-grouper is used, at least most the code seems unused

This code is being used. The only method used is `CodeGrouper.reduce(answer, base, language)`. That gem could use some cleanup, its very old, and obviously way out of date. Basic comment stripping is only implemented for 3 languages. I guess the basic syntax stripping has done a decent enough job that the code grouping hasn't been called out. I completely forgot to update this to be honest. 

If I recall, the Levenshtein distance code ended up being not used due to some issues with it falsely grouping some things. I think I went back and forth on it, commented it out due to some issues, got sidetracked with a million different other things and never revisited this. 

I would love some help if anyone is up for it.

**Comment by kazk** [comment]
>  I guess the basic syntax stripping has done a decent enough job that the code grouping hasn't been called out.

I actually saw an edge case recently. JavaScript groups these together (modified a little from actual solutions):

```
function makePizza(pieces) {        | function makePizza(pieces) {
  var result = "";                  |   var result = "";
  while (pieces--) {                |   while (pieces--) {
    var needSteps = 5;              |     var needSteps = 5;
    while (needSteps--)             |     while (needSteps--)
      result += make(needSteps+1);  |       result += make(needSteps);
  }                                 |   }
  return result;                    |   return result;
}                                   | }
function make(step) {               | function make(step) {
  switch (step) {                   |   switch (step+1) {
    case 5: return "P";             |     case 5: return "P";
    case 4: return "i";             |     case 4: return "i";
    case 3:                         |     case 3:
    case 2: return "z";             |     case 2: return "z";
    case 1: return "a";             |     case 1: return "a";
  }                                 |   }
}                                   | }
```

The left has `+1` in `makePizza`, while the right has it in `make`.

---

> If I recall, the Levenshtein distance code ended up being not used due to some issues with it falsely grouping some things.

I wouldn't use Levenshtein distance (or any edit distance measures) to group codes because single character makes a huge difference.

> I would love some help if anyone is up for it.

Does it have to be entirely in Ruby? For JavaScript, I found [danielstjules/jsinspect](https://github.com/danielstjules/jsinspect).

**Comment by kazk** [comment]
Found a Ruby gem [flay](https://rubygems.org/gems/flay) used by [Code Climate for code duplication analysis](https://github.com/codeclimate/codeclimate-duplication).

**Comment by jhoffner** [comment]
> Does it have to be entirely in Ruby? For JavaScript, I found danielstjules/jsinspect.

Yes. Our servers can only run Ruby. I guess we could in theory run JS through an embedded engine but that's going to probably cause memory issues. 

Both of the packages you mentioned are unfortunately specific to JS or Ruby. Another option would be to start adding language specific packages to the Coderunner CLI and then call out to that. This would also eventually open up to allowing static code analysis. 

One thing to keep in mind with code de-duplication, is that many of the libraries that I had looked at in the past tried to use abstract syntax trees to compare code. Thats not great for our use case because even though 2 sets of solutions might be fundamentally the same thing to the compiler/interpreter, they may not be the same thing in terms of which is a better practice. It would suck to have hard to read code golfed solutions mixed in with clean code.

**Comment by kazk** [comment]
> Another option would be to start adding language specific packages to the Coderunner CLI and then call out to that. This would also eventually open up to allowing static code analysis.

I like this, but it'll be a lot of work.

> It would suck to have hard to read code golfed solutions mixed in with clean code.

Ok, now I understand why you used a string based approach :) I didn't think about that.
I'll let you know if I find anything useful.

**Comment by anter69** [comment]
So, extra honor is out of the question -- fine.

By trying to read the code, I believe basic python support (comment stripping) could be added easily by updating the relevant regex.

On the other hand, reading the readme of code-grouper, I was wondering why the example `function sortList (sortBy, list)` should **not** be grouped with `function sortList (s, l)` and how it is decided.

And finally: is any  check done on python code, or it's skipped altogether?

**Comment by jhoffner** [comment]
> I was wondering why the example `function sortList (sortBy, list)` should not be grouped with `function sortList (s, l)` and how it is decided.

Because `sortBy` and `list` are great parameter names and `s` and `l` are not.

**Comment by leerssej** [comment]
Would it be possible to group the solutions employing similar approaches together, as well as generate a metric to quantify use of descriptive variable names versus minified names? (Thinking something similar to the way that the `reduced_difference` is being calculated.) 
Then you can have cohesive taxa, as well as being able to rank their members along an axis of readability.
